File: fiberplane-charts/src/Mondrian/generator/utils/calculateBarWidth.ts
import type { Axis } from "../../types";
import { BAR_PLUS_PADDING } from "../constants";

/**
 * Calculates the width of bars in bar charts.
 */
export function calculateBarWidth(
  xAxis: Axis,
  interval: number,
  numBarsPerGroup: number,
): number {
  const numGroups =
    interval === 0
      ? 1
      : Math.round((xAxis.maxValue - xAxis.minValue) / interval) + 1;
  const numBars = numGroups * numBarsPerGroup;
  return 1 / (numBars * BAR_PLUS_PADDING);
}

File: mondrian-charts/src/utils/calculate_bar_width.rs
use crate::constants::BAR_PLUS_PADDING;
use crate::types::Axis;

/// Calculates the width of bars in bar charts.
pub(crate) fn calculate_bar_width(
    x_axis: &Axis,
    interval: Option<f64>,
    num_bars_per_group: usize,
) -> f64 {
    let num_groups = if let Some(interval) = interval {
        ((x_axis.max_value - x_axis.min_value) / interval).round() + 1.
    } else {
        1.
    };
    let num_bars = num_groups * num_bars_per_group as f64;
    1. / (num_bars * BAR_PLUS_PADDING)
}
